protocols/kad/src/record/store.rs

```diff
@@ -0,0 +1,95 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+mod memory;
+
+pub use memory::MemoryStore;
+
+use crate::K_VALUE;
+use super::*;
+use std::borrow::Cow;
+
+/// The result of an operation on a `RecordStore`.
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// The possible errors of a `RecordStore` operation.
+#[derive(Debug)]
+pub enum Error {
+    /// The store is at capacity w.r.t. the total number of stored records.
+    MaxRecords,
+    /// The store is at capacity w.r.t. the total number of stored keys for
+    /// provider records.
+    MaxProvidedKeys,
+    /// The value of a record to be stored is too large.
+    ValueTooLarge,
+}
+
+/// Trait for types implementing a record store.
+///
+/// There are two types of records managed by a `RecordStore`:
+///
+///   1. Regular (value-)records. These records store an arbitrary value
+///      associated with a key which is distributed to the closest nodes
+///      to the key in the Kademlia DHT as per the standard Kademlia "push-model".
+///      These records are subject to re-replication and re-publication as
+///      per the standard Kademlia protocol.
+///
+///   2. Provider records. These records associate the ID of a peer with a key
+///      who can supposedly provide the associated value. These records are
+///      mere "pointers" to the data which may be followed by contacting these
+///      providers to obtain the value. These records are specific to the
+///      libp2p Kademlia specification and realise a "pull-model" for distributed
+///      content. Just like a regular record, a provider record is distributed
+///      to the closest nodes to the key.
+///
+pub trait RecordStore<'a> {
+    type RecordsIter: Iterator<Item = Cow<'a, Record>>;
+    type ProvidedIter: Iterator<Item = Cow<'a, ProviderRecord>>;
+
+    /// Gets a record from the store, given its key.
+    fn get(&'a self, k: &Multihash) -> Option<Cow<Record>>;
```

Comment:
Should eventually be asynchronous, but this would probably be extremely hard to implement as long as we don't have async/await, and not worth it.

Reply:
I thought the same - besides we have https://github.com/libp2p/rust-libp2p/issues/1159 - but I am myself not sure whether the complications w.r.t. the implementation of the Kademlia behaviour that would come along with such a change would be worth it, especially not before async/await. I think a first step would be to actually allow even `RecordStore` implementations using synchronous I/O without `unwrap`ing, which will require all trait methods to return a `Result` and `store::Error` would need a constructor for `io::Error`. I left that for later as well.